### PR TITLE
Supporting decorators in new views-based Search.

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

This PR is adding support for backend decorators to the new views-based search. In the existing search, backend decorators are added by someone with the required permission (`decorators:create`/`decorators:edit`/`decorators:read`) per stream, for which they are then added to all searches started for that stream for every user. They are effectively shared globally between all users, with some being able to modify them. The set of decorators created for a given stream are references as "default decorators" from now on.

For the views this concept seemed to be too unflexible. Instead, we are now retrieving the current set of decorators defined for the current stream (if a new search is started through the `Streams` page) or the default set (if a new search is started by clicking on `Search`) and added to the message table which is generated for a new search automatically. Decorator sets now function as templates, which can be modified by the search author. In addition, if no default decorators exist, a view author can add new decorators to an own search without owning any special permission by editing any message widget.

Default decorators can now be managed by anyone with the required permissions on the `System / Configurations` page, where a separate config component exists, which shows existing decorators, allows updating, creating and removing decorators.

The fact that a field was decorated, is shown in the message details by adding a small marker next to the field's value. Value actions are disabled for this value, because it is not a materialized value, so most actions will not return the expected results.

## Screenshots (if appropriate):
Screen Shot 2019-12-11 at 15 59 30
Screen Shot 2019-12-11 at 16 03 37
Screen Shot 2019-12-11 at 16 03 51
Screen Shot 2019-12-11 at 16 41 46

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by kmerz** [comment]
Graylog (35)

Cropped drop down

**Comment by kmerz** [comment]
Decorated Fields can't be aggregated and therefore field actions should disable that:
Graylog (36)

**Comment by kmerz** [comment]
Graylog (37)
My pipeline decorator get cut before being useful.

**Comment by kmerz** [comment]
Hyperlink decorator do not seem to work 
Graylog (37)

**Comment by kmerz** [comment]
Hyperlink decorator do not seem to work:
Graylog (38)

**Comment by kmerz** [comment]
When clicking on the configurations page I only get: 

`DecoratorsConfig.jsx:51 Uncaught TypeError: Cannot read property 'sort' of undefined
    at DecoratorsConfig (DecoratorsConfig.jsx:51)`

File: graylog2-web-interface/src/views/logic/views/ViewStateGenerator.js [hunk 1]
```diff
@@ -30,7 +35,8 @@ const _defaultWidgets = {
 
     return { titles, widgets, positions };
   },
-  [View.Type.Dashboard]: () => {
+  // eslint-disable-next-line no-unused-vars
+  [View.Type.Dashboard]: async (streamId: ?string) => {
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/logic/views/ViewStateGenerator.js]: Yes, that is the signature that we expect from all functions in there.

File: graylog2-web-interface/src/components/configurations/DecoratorsConfig.jsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                              streams={streams}
+                              decorators={decorators}
+                              onCancel={closeModal}
+                              onSave={onSave}
+                              types={types} />
+    </div>
+  );
+};
+
+DecoratorsConfig.propTypes = {};
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/components/configurations/DecoratorsConfig.jsx]: It does not really matter. In this case it was part of my template which I used to create a new component.

File: graylog2-web-interface/src/views/components/messagelist/decorators/DecoratorList.jsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    onReorder(...args);
+  };
+
+  render() {
+    const { decorators, disableDragging } = this.props;
+    if (!decorators || decorators.length === 0) {
+      return (
+        <Alert bsStyle="info" className={DecoratorStyles.noDecoratorsAlert}>
+          <Icon name="info-circle" />&nbsp;No decorators configured.
+        </Alert>
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/messagelist/decorators/DecoratorList.jsx]: ✅

File: graylog2-web-interface/src/components/configurations/decorators/DecoratorsConfigUpdate.jsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import { Button, Modal } from 'components/graylog';
+import { IfPermitted } from 'components/common';
+import DecoratorList from 'views/components/messagelist/decorators/DecoratorList';
+import AddDecoratorButton from 'views/components/messagelist/decorators/AddDecoratorButton';
+import type { Decorator } from 'views/components/messagelist/decorators/Types';
+import StreamSelect, { DEFAULT_SEARCH_ID, DEFAULT_STREAM_ID } from './StreamSelect';
+import formatDecorator from './FormatDecorator';
+
+type Props = {
+  streams: Array<{ title: string, id: string }>,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/components/configurations/decorators/DecoratorsConfigUpdate.jsx]: ✅

File: graylog2-web-interface/src/views/components/visualizations/bar/BarVisualization.jsx [hunk 5]
```diff
@@ -62,7 +62,7 @@ const BarVisualization: VisualizationComponent = ({ config, data, effectiveTimer
 
 BarVisualization.propTypes = {
   config: AggregationType.isRequired,
-  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+  data: PropTypes.object.isRequired,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/visualizations/bar/BarVisualization.jsx]: This is a quick fix, that is at least improving the situation. We should use a better type in the future though, describing the structure better (or at all).

File: graylog2-web-interface/src/components/common/Select.jsx [hunk 6]
```diff
@@ -166,7 +166,6 @@ type Props = {
   value?: string,
   autoFocus?: boolean,
   size?: 'normal' | 'small',
-  optionRenderer: (any) => React.Node,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/components/common/Select.jsx]: 👍

**Pull request merged**